server/legalhold/legal_hold.go

```diff
@@ -40,42 +46,84 @@ type Execution struct {
 }
 
 // NewExecution creates a new Execution that is ready to use.
-func NewExecution(legalHold model.LegalHold, papi plugin.API, store *sqlstore.SQLStore, fileBackend filestore.FileBackend) Execution {
+func NewExecution(legalHold model.LegalHold, papi plugin.API, store *sqlstore.SQLStore, kvstore kvstore.KVStore, fileBackend filestore.FileBackend) Execution {
 	return Execution{
 		LegalHold:          legalHold,
 		ExecutionStartTime: legalHold.NextExecutionStartTime(),
 		ExecutionEndTime:   legalHold.NextExecutionEndTime(),
 		store:              store,
+		kvstore:            kvstore,
 		fileBackend:        fileBackend,
 		index:              model.NewLegalHoldIndex(),
 		papi:               papi,
 		hashes:             make(map[string]string),
 	}
 }
 
-// Execute executes the Execution.
-func (ex *Execution) Execute() (int64, error) {
-	err := ex.GetChannels()
+// Execute executes the Execution and returns the updated LegalHold.
+func (ex *Execution) Execute() (*model.LegalHold, error) {
+	now := time.Now().Unix()
+
+	// Lock multiple executions using a cluster mutex
+	mutexKey := fmt.Sprintf("legal_hold_%s_execution", ex.LegalHold.ID)
```

Comment:
I'm thinking we need to limit the number of running legal hold executions to avoid overloading the server.  With the current named mutex including the legal hold id we can have every legal hold running at the same time.  One legal hold at a time would be fine IMO, or one job and one manual run.

Reply:
Let's use one legal hold for now.

Reply:
https://github.com/mattermost/mattermost-plugin-legal-hold/pull/129/commits/5e44cab716f76a0bec43ea9a99540a9e69455d93